File: ios/OTPPoc/OTPPoc/Networking/APIClient.swift
import Foundation

enum APIError: LocalizedError {
    case serverError(code: String, message: String, retryAfter: Int?)
    case networkError(Error)
    case decodingError(Error)

    var errorDescription: String? {
        switch self {
        case .serverError(let code, let message, _):
            return "\(code): \(message)"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .decodingError(let error):
            return "Decoding error: \(error.localizedDescription)"
        }
    }
}

final class APIClient {
    static let shared = APIClient()

    // iOS simulator shares host network — localhost works
    private let baseURL = URL(string: "http://localhost:3000")!
    private let session = URLSession.shared
    private let tokenManager = TokenManager.shared
    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = .sortedKeys
        return e
    }()
    private let decoder = JSONDecoder()

    func request<T: Decodable>(
        method: String,
        path: String,
        body: (any Encodable)? = nil,
        extraHeaders: [String: String] = [:]
    ) async throws -> T {
        let url = baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(tokenManager.deviceId, forHTTPHeaderField: "X-Device-Id")
        request.setValue("mobile", forHTTPHeaderField: "X-Channel")

        // Body
        var bodyData: Data?
        if let body {
            bodyData = try encoder.encode(body)
            request.httpBody = bodyData
        }

        // HMAC signing
        let (signature, timestamp, nonce) = HMACSigner.sign(
            method: method,
            path: path,
            body: bodyData
        )
        request.setValue(signature, forHTTPHeaderField: "X-Signature")
        request.setValue(timestamp, forHTTPHeaderField: "X-Timestamp")
        request.setValue(nonce, forHTTPHeaderField: "X-Nonce")

        // Auth token
        if let accessToken = tokenManager.accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }

        // Attestation token for mobile
        if let attestJwt = tokenManager.attestationJwt {
            request.setValue(attestJwt, forHTTPHeaderField: "X-Attestation-Token")
        }

        // Extra headers
        for (key, value) in extraHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIError.networkError(URLError(.badServerResponse))
        }

        let apiResponse: ApiResponse<T>
        do {
            apiResponse = try decoder.decode(ApiResponse<T>.self, from: data)
        } catch {
            throw APIError.decodingError(error)
        }

        if !apiResponse.success, let error = apiResponse.error {
            throw APIError.serverError(
                code: error.code,
                message: error.message,
                retryAfter: error.retry_after
            )
        }

        guard let result = apiResponse.data else {
            throw APIError.serverError(code: "NO_DATA", message: "No data in response", retryAfter: nil)
        }

        return result
    }

    // Convenience methods

    func preflight(clientSignals: ClientSignals? = nil) async throws -> PreflightResponse {
        let body = PreflightRequest(
            channel: "mobile",
            device_id: tokenManager.deviceId,
            fingerprint: nil,
            client_signals: clientSignals
        )
        return try await request(method: "POST", path: "/v1/auth/preflight", body: body)
    }

    func sendOtp(
        phone: String,
        purpose: String,
        preflightToken: String,
        powSolution: PowSolutionRequest? = nil,
        captchaToken: String? = nil
    ) async throws -> OtpSendResponse {
        let body = OtpSendRequest(
            phone: phone,
            purpose: purpose,
            pow_solution: powSolution,
            captcha_token: captchaToken
        )
        return try await request(
            method: "POST",
            path: "/v1/auth/otp/send",
            body: body,
            extraHeaders: ["X-Preflight-Token": preflightToken]
        )
    }

    func verifyOtp(phone: String, code: String, challengeId: String) async throws -> OtpVerifyResponse {
        let body = OtpVerifyRequest(
            phone: phone,
            code: code,
            challenge_id: challengeId,
            device_id: tokenManager.deviceId
        )
        return try await request(method: "POST", path: "/v1/auth/otp/verify", body: body)
    }

    func getSession() async throws -> SessionMeResponse {
        return try await request(method: "GET", path: "/v1/auth/session/me")
    }
}

File: android/app/src/main/java/com/otppoc/android/networking/ApiClient.kt
package com.otppoc.android.networking

import android.content.Context
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.otppoc.android.models.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.TreeMap
import java.util.concurrent.TimeUnit

class ApiClient private constructor(context: Context) {

    // Android emulator uses 10.0.2.2 to reach host machine
    private val baseUrl = "http://10.0.2.2:3000"
    private val tokenManager = TokenManager.getInstance(context)
    private val gson: Gson = GsonBuilder().serializeNulls().create()
    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .writeTimeout(30, TimeUnit.SECONDS)
        .build()

    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

    /**
     * Recursively sorts all map keys to ensure deterministic JSON for HMAC body hash.
     */
    @Suppress("UNCHECKED_CAST")
    private fun sortKeys(obj: Any?): Any? {
        return when (obj) {
            is Map<*, *> -> {
                val sorted = TreeMap<String, Any?>()
                for ((key, value) in obj) {
                    sorted[key as String] = sortKeys(value)
                }
                sorted
            }
            is List<*> -> obj.map { sortKeys(it) }
            else -> obj
        }
    }

    private fun toSortedJson(body: Any): String {
        // Serialize to a generic map first, then sort recursively, then serialize back
        val jsonString = gson.toJson(body)
        val mapType = object : TypeToken<Map<String, Any?>>() {}.type
        val unsorted: Map<String, Any?> = gson.fromJson(jsonString, mapType)
        val sorted = sortKeys(unsorted)
        return gson.toJson(sorted)
    }

    suspend fun <T> request(
        method: String,
        path: String,
        body: Any? = null,
        extraHeaders: Map<String, String> = emptyMap(),
        responseType: TypeToken<ApiResponse<T>>
    ): T = withContext(Dispatchers.IO) {
        val url = "$baseUrl$path"

        // Build body
        var bodyBytes: ByteArray? = null
        val requestBuilder = Request.Builder().url(url)

        if (body != null) {
            val jsonBody = toSortedJson(body)
            bodyBytes = jsonBody.toByteArray(Charsets.UTF_8)
            requestBuilder.method(method, bodyBytes.toRequestBody(jsonMediaType))
        } else {
            when (method) {
                "POST" -> requestBuilder.method(method, "".toRequestBody(jsonMediaType))
                else -> requestBuilder.method(method, null)
            }
        }

        // Standard headers
        requestBuilder.addHeader("Content-Type", "application/json")
        requestBuilder.addHeader("X-Device-Id", tokenManager.deviceId)
        requestBuilder.addHeader("X-Channel", "mobile")

        // HMAC signing
        val sig = HmacSigner.sign(method, path, bodyBytes)
        requestBuilder.addHeader("X-Signature", sig.signature)
        requestBuilder.addHeader("X-Timestamp", sig.timestamp)
        requestBuilder.addHeader("X-Nonce", sig.nonce)

        // Auth token
        tokenManager.accessToken?.let {
            requestBuilder.addHeader("Authorization", "Bearer $it")
        }

        // Attestation token
        tokenManager.attestationJwt?.let {
            requestBuilder.addHeader("X-Attestation-Token", it)
        }

        // Extra headers
        for ((key, value) in extraHeaders) {
            requestBuilder.addHeader(key, value)
        }

        val request = requestBuilder.build()
        val response = client.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw ApiException("NO_DATA", "Empty response body")

        val apiResponse: ApiResponse<T> = gson.fromJson(responseBody, responseType.type)

        if (!apiResponse.success || apiResponse.error != null) {
            val err = apiResponse.error
            throw ApiException(
                code = err?.code ?: "UNKNOWN",
                message = err?.message ?: "Unknown error",
                retryAfter = err?.retry_after
            )
        }

        apiResponse.data ?: throw ApiException("NO_DATA", "No data in response")
    }

    // Convenience methods

    suspend fun preflight(clientSignals: ClientSignals? = null): PreflightResponse {
        val body = PreflightRequest(
            channel = "mobile",
            device_id = tokenManager.deviceId,
            fingerprint = null,
            client_signals = clientSignals
        )
        return request(
            method = "POST",
            path = "/v1/auth/preflight",
            body = body,
            responseType = object : TypeToken<ApiResponse<PreflightResponse>>() {}
        )
    }

    suspend fun sendOtp(
        phone: String,
        purpose: String,
        preflightToken: String,
        powSolution: PowSolutionRequest? = null,
        captchaToken: String? = null
    ): OtpSendResponse {
        val body = OtpSendRequest(
            phone = phone,
            purpose = purpose,
            pow_solution = powSolution,
            captcha_token = captchaToken
        )
        return request(
            method = "POST",
            path = "/v1/auth/otp/send",
            body = body,
            extraHeaders = mapOf("X-Preflight-Token" to preflightToken),
            responseType = object : TypeToken<ApiResponse<OtpSendResponse>>() {}
        )
    }

    suspend fun verifyOtp(phone: String, code: String, challengeId: String): OtpVerifyResponse {
        val body = OtpVerifyRequest(
            phone = phone,
            code = code,
            challenge_id = challengeId,
            device_id = tokenManager.deviceId
        )
        return request(
            method = "POST",
            path = "/v1/auth/otp/verify",
            body = body,
            responseType = object : TypeToken<ApiResponse<OtpVerifyResponse>>() {}
        )
    }

    suspend fun getSession(): SessionMeResponse {
        return request(
            method = "GET",
            path = "/v1/auth/session/me",
            responseType = object : TypeToken<ApiResponse<SessionMeResponse>>() {}
        )
    }

    suspend fun attestChallenge(): AttestationChallengeResponse {
        val body = mapOf("device_id" to tokenManager.deviceId)
        return request(
            method = "POST",
            path = "/v1/device/attest",
            body = body,
            responseType = object : TypeToken<ApiResponse<AttestationChallengeResponse>>() {}
        )
    }

    suspend fun attestVerify(body: AttestationVerifyRequest): AttestationVerifyResponse {
        return request(
            method = "POST",
            path = "/v1/device/attest/verify",
            body = body,
            responseType = object : TypeToken<ApiResponse<AttestationVerifyResponse>>() {}
        )
    }

    companion object {
        @Volatile
        private var instance: ApiClient? = null

        fun getInstance(context: Context): ApiClient {
            return instance ?: synchronized(this) {
                instance ?: ApiClient(context.applicationContext).also { instance = it }
            }
        }
    }
}

class ApiException(
    val code: String,
    override val message: String,
    val retryAfter: Int? = null
) : Exception("$code: $message")
